cis194/week7/zehua/src/JoinList.hs

```diff
@@ -0,0 +1,174 @@
+module JoinList where
+
+
+import           Scrabble
+import           Sized
+
+data JoinList m a = Empty
+                  | Single m a
+                  | Append m (JoinList m a) (JoinList m a)
+  deriving (Eq, Show)
+
+-- ex1
+tag :: Monoid m => JoinList m a -> m
+tag Empty          = mempty
+tag (Single m _)   = m
+tag (Append m _ _) = m
+
+(+++) :: Monoid m => JoinList m a -> JoinList m a -> JoinList m a
+Empty +++ a     = a
+a     +++ Empty = a
+j1    +++ j2    = Append (m1 <> m2) j1 j2
+  where
+    m1 = tag j1
+    m2 = tag j2
+
+
+-- ex2
+sizedTag :: (Sized m, Monoid m) => JoinList m a -> m
+sizedTag = tag
+
+getSizeFromTag :: (Sized m, Monoid m) => JoinList m a -> Int
+getSizeFromTag = getSize . size . sizedTag
+
+foldJoinList :: (Sized b, Monoid b) =>
```

Comment:
the fold version seems overly complicated. I included them here just as an exercise.